rcl/src/rcl/qos_compatibility.c

```diff
@@ -0,0 +1,78 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rcl/qos_compatibility.h"
+
+#include "rcl/error_handling.h"
+#include "rcl/graph.h"
+#include "rcl/types.h"
+
+#include "rmw/qos_profiles.h"
+
+rcl_ret_t
+rcl_get_compatible_qos_for_topic_subscription(
+  const rcl_node_t * node,
+  const char * topic_name,
+  rmw_qos_profile_t * subscription_qos_profile)
+{
+  if (!rcl_node_is_valid(node)) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(topic_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(subscription_qos_profile, RCL_RET_INVALID_ARGUMENT);
+
+  const rcl_node_options_t * node_options = rcl_node_get_options(node);
+  const rcl_allocator_t * allocator = &node_options->allocator;
+
+  rcl_topic_endpoint_info_array_t publishers_info =
+    rcl_get_zero_initialized_topic_endpoint_info_array();
+  rcl_ret_t ret = rcl_get_publishers_info_by_topic(
+    node,
+    // TODO(jacobperron): Update rcl_get_publishers_info_by_topic API to take an rcl_allocator_t 
+    (rcutils_allocator_t *)allocator,
+    topic_name,
+    false,
+    &publishers_info);
+  if (RCL_RET_OK != ret) {
+    return ret;
+  }
+
+  const size_t number_of_publishers = publishers_info.size;
+  if (0u == number_of_publishers) {
+    return RCL_RET_OK;
+  }
+
+  // Copy QoS profiles to array
+  rmw_qos_profile_t * publisher_qos_profiles = (rmw_qos_profile_t *)allocator->allocate(
+    number_of_publishers * sizeof(rmw_qos_profile_t),
+    allocator->state);
+  for (size_t i = 0u; i < number_of_publishers; ++i) {
+    publisher_qos_profiles[i] = publishers_info.info_array[i].qos_profile;
+  }
+
+  rmw_ret_t rmw_ret = rmw_qos_profile_get_most_compatible_for_subscription(
+    publisher_qos_profiles,
```

Comment:
This is just an idea, but what if instead this first parameter was of type `rcl_topic_endpoint_info_array_t` (same as `rmw_topic_endpoint_info_array_t`)? I guess if you're not using `rcl_get_publishers_info_by_topic()` (e.g. during testing), it's slightly harder to use, but it would simplify this function considerably.

Do you think this `rmw_qos_profile_get_most_compatible_for_subscription()` will often be used outside of this situation where the qos profiles you're matching would be coming from another source?

I'm not totally convinced myself, just wanted to bring it up.

Reply:
I don't know have any situations in mind where `rmw_qos_profile_get_most_compatible_for_subscription` would be used other than like this, so your point it valid. It might actually be nicer, since we could return an array of endpoint GIDs for any that are incompatible instead of an array of bool.